### PR TITLE
Improve the impl to restrict to add field annotation without entity annotation

### PR DESCRIPTION
**Pull request by kalaiyarasiganeshalingam**
## Purpose
$Subject

Fixes:https://github.com/ballerina-platform/ballerina-standard-library/issues/3683

## Checklist
- [ ] Linked to an issue
- [ ] Updated the specification
- [ ] Updated the changelog
- [ ] Added tests

### CODE REVIEW

File: compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/DiagnosticsCodes.java [hunk 1]
```diff
@@ -74,6 +74,10 @@ public enum DiagnosticsCodes {
     PERSIST_123("PERSIST_123", "associated entity does not contain any primary keys: the key " +
             "should have a valid value", ERROR),
     PERSIST_124("PERSIST_124", "The entity ''{0}'' should be a closed record.", ERROR),
+    PERSIST_125("PERSIST_125", "invalid annotation attachment: relation annotation cannot be " +
```
Comment (niveathika) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/DiagnosticsCodes.java:77]:
```suggestion
    PERSIST_125("PERSIST_125", "relation annotation cannot be " +
```

File: compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/PersistRecordValidator.java [hunk 2]
```diff
@@ -137,8 +120,8 @@ public void perform(SyntaxNodeAnalysisContext ctx) {
                     validateEntityAnnotation(ctx, typeDefinitionNode, recordTypeSymbol);
                     validateRecordFieldsAnnotation(ctx, recordNode,
                             ((ModulePartNode) ctx.syntaxTree().rootNode()).members());
-                    validateRecordField(ctx, typeDefinitionNode, recordTypeSymbol, memberNodes);
-                    if ((hasPersistAnnotation || isPersistEntity)) {
+                    if (hasPersistAnnotation) {
+                        validateRecordFieldType(ctx, recordTypeSymbol.fieldDescriptors());
```
Comment (niveathika) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/PersistRecordValidator.java:124]: Wouldnt it be easier to check this case in the else part of if clause in 123. This is much clearer. If you dont have persist annotation, we just have to loop through field with annotation and add error if they use other annotations.

File: compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/DiagnosticsCodes.java [hunk 3]
```diff
@@ -58,7 +58,7 @@ public enum DiagnosticsCodes {
             "does not have the field with the relationship type", ERROR),
     PERSIST_116("PERSIST_116", "invalid entity initialisation: the relation annotation should " +
             "only be added to the relationship owner for one-to-one and one-to-many associations", ERROR),
-    PERSIST_117("PERSIST_117", "invalid annotation attachment: The relation annotation can only be " +
+    PERSIST_117("PERSIST_117", "invalid annotation attachment: the relation annotation can only be " +
             "attached to the entity record field", ERROR),
```
Comment (kaneeldias) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/DiagnosticsCodes.java:62]:
```suggestion
            "attached to an entity record field", ERROR),
```

File: compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/DiagnosticsCodes.java [hunk 4]
```diff
@@ -74,6 +74,10 @@ public enum DiagnosticsCodes {
     PERSIST_123("PERSIST_123", "associated entity does not contain any primary keys: the key " +
             "should have a valid value", ERROR),
     PERSIST_124("PERSIST_124", "The entity ''{0}'' should be a closed record.", ERROR),
+    PERSIST_125("PERSIST_125", "relation annotation cannot be attached without entity annotation",
+            ERROR),
+    PERSIST_126("PERSIST_126", "auto increment annotation cannot be attached without entity annotation",
+            ERROR),
```
Comment (kaneeldias) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/DiagnosticsCodes.java:80]:
```suggestion
    PERSIST_125("PERSIST_125", "relation annotation cannot be attached without the entity annotation",
            ERROR),
    PERSIST_126("PERSIST_126", "auto increment annotation cannot be attached without the entity annotation",
            ERROR),
```

File: compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/PersistRecordValidator.java [hunk 5]
```diff
@@ -264,30 +178,66 @@ private void validateRecordFieldsAnnotation(SyntaxNodeAnalysisContext ctx, Node
...
         if (!isReadOnly) {
             reportDiagnosticInfo(ctx, location, DiagnosticsCodes.PERSIST_106.getCode(),
                     DiagnosticsCodes.PERSIST_106.getMessage(), DiagnosticsCodes.PERSIST_106.getSeverity());
         }
     }
 
-    private void validateAnnotation(SyntaxNodeAnalysisContext ctx, MetadataNode metadataNode, String filedType,
-                                    Location location, NodeList<ModuleMemberDeclarationNode> memberNodes,
-                                    String fieldName) {
+    private void validateAnnotationFields(SyntaxNodeAnalysisContext ctx, MetadataNode metadataNode, String filedType,
```
Comment (kaneeldias) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/PersistRecordValidator.java:238]:
```suggestion
    private void validateAnnotationFields(SyntaxNodeAnalysisContext ctx, MetadataNode metadataNode, String fieldType,
```

File: compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/PersistRecordValidator.java [hunk 6]
```diff
@@ -264,30 +178,66 @@ private void validateRecordFieldsAnnotation(SyntaxNodeAnalysisContext ctx, Node
...
                 Optional<MetadataNode> metadataNode = fieldNode.metadata();
-                metadataNode.ifPresent(node -> validateAnnotation(ctx, node, fieldNode.typeName().toSourceCode().trim(),
+                metadataNode.ifPresent(node -> validateAnnotationFields(ctx, node,
+                        fieldNode.typeName().toSourceCode().trim(),
                         fieldNode.location(), memberNodes, fieldNode.fieldName().text().trim()));
             }
         }
     }
 
+    private void checkRecordFieldsAnnotation(SyntaxNodeAnalysisContext ctx, Node recordNode) {
```
Comment (kaneeldias) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/PersistRecordValidator.java:197]: This method has nearly the same functionality as `validateRecordFieldsAnnotation()`?

**Review by kaneeldias — DISMISSED** [review]

**Comment by kaneeldias** [comment]
Other than 2 minor comments LGTM

File: compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/PersistRecordValidator.java (hunk 6 above)
Comment (kalaiyarasiganeshalingam) [reply compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/PersistRecordValidator.java]: Updated to `checkFieldAnnotationPresent`

**Review by kaneeldias — APPROVED** [review]

File: compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/DiagnosticsCodes.java [hunk 7]
```diff
@@ -74,6 +74,10 @@ public enum DiagnosticsCodes {
     PERSIST_123("PERSIST_123", "associated entity does not contain any primary keys: the key " +
             "should have a valid value", ERROR),
     PERSIST_124("PERSIST_124", "The entity ''{0}'' should be a closed record.", ERROR),
+    PERSIST_125("PERSIST_125", "relation annotation cannot be attached without the entity annotation",
```
Comment (niveathika) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/DiagnosticsCodes.java:77]: Whats the difference between 117 amd 125?

Comment (niveathika) [reply compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/DiagnosticsCodes.java:77]: I prefer the wording in 117

File: compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/PersistRecordValidator.java [hunk 8]
```diff
@@ -126,24 +111,24 @@ public void perform(SyntaxNodeAnalysisContext ctx) {
...
+                        validateRecordFieldsAnnotation(ctx, recordNode,
+                                ((ModulePartNode) ctx.syntaxTree().rootNode()).members());
+                        validateRecordFieldType(ctx, recordTypeSymbol.fieldDescriptors());
                         validateRecordType(ctx, typeDefinitionNode);
                         if (this.noOfReportDiagnostic == 0) {
                             validFieldTypeAndRelation((RecordTypeDescriptorNode) recordNode, typeDefinitionNode,
                                     ctx, symbol.get());
                         }
+                    } else {
+                        checkFieldAnnotationPresent(ctx, recordNode);
```
Comment (niveathika) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/PersistRecordValidator.java:131]:
```suggestion
                        checkPresenceOfFieldAnnotation(ctx, recordNode);
```

File: compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/PersistRecordValidator.java [hunk 9]
```diff
@@ -264,30 +178,66 @@ private void validateRecordFieldsAnnotation(SyntaxNodeAnalysisContext ctx, Node
...
+            }
+        }
+    }
+
+    private void validateAnnotation(SyntaxNodeAnalysisContext ctx, MetadataNode metadataNode, Location location) {
+        NodeList<AnnotationNode> annotations = metadataNode.annotations();
+        for (AnnotationNode annotation : annotations) {
+            String annotationName = annotation.annotReference().toSourceCode().trim();
+            if (annotationName.equals(Constants.AUTO_INCREMENT)) {
+                if (!hasPersistAnnotation) {
```
Comment (niveathika) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/PersistRecordValidator.java:218]: Why is this check repeated? Isn't it done at the first instance itself

File: compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/DiagnosticsCodes.java (hunk 7 above)
Comment (kalaiyarasiganeshalingam) [reply compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/DiagnosticsCodes.java:77]: We can't add the relation annotation if the field type is not another entity field type. 117 is created to this validation.
Eg: 
```
public type MultipleAssociations record {|
    readonly int id;
    @persist:Relation
    string name;
}
```

File: compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/PersistRecordValidator.java (hunk 9 above)
Comment (kalaiyarasiganeshalingam) [reply compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/PersistRecordValidator.java:218]: Fixed in #73

File: compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/DiagnosticsCodes.java (hunk 7 above)
Comment (niveathika) [reply compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/DiagnosticsCodes.java:77]: Lets word this differently. IMO both messages are more or less the same and I thought it meant the same thing. We have to be clear on messaging

**Pull request merged**